core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala

```diff
@@ -433,11 +445,13 @@ private[zio] final class FiberContext[E, A](
       null
     }
 
-  private[this] final val exitUninterruptible: UIO[Unit] = IO.effectTotal { noInterrupt -= 1 }
+  private[this] final def noInterrupt[E, A](io: IO[E, A]): IO[E, A] =
+    changeInterrupt(io, false)
 
-  private[this] final def doNotInterrupt[E, A](io: IO[E, A]): IO[E, A] = {
-    this.noInterrupt += 1
-    io.ensuring(exitUninterruptible)
+  private[this] final def changeInterrupt[E, A](io: IO[E, A], flag: Boolean): IO[E, A] = {
+    interruptStatus.push(flag)
```

Comment:
Won't it make `foo.interruptible.uninterruptible.uninterruptible` and, more importantly, `bar.ensuring(foo.interruptible)` interruptible in `foo`? Do you really want that? Basically, you can make any piece of code *forcibly* interruptible in any context now.

Reply:
If the nested effect is uninterruptible, then it cannot be made uninterruptible. 

`interruptible` only creates a "hole" in which the outer `uninterruptible` does not apply; however, inside that hole, if there are composed effects that are `uninterruptible`, then they will remain `uninterruptible`. Indeed, tracking all this information (on monadic region) is the main point of `StackBool`.

Reply:
As far as i can tell, it does not create a "hole" in the outer `uninterruptible` (that would be peeling one element off from the stack), it forces interruptibility to `true`. That is not composable. For example, if you have a function
```scala
def foo = (... bar.interruptible ...).uninterruptible
```
then that's it, you can *never* make that effect run to its end reliably (e.g. in `bracket`'s `release`). `interruptible` should restore interruptibility to the value it had before the outer `uninterruptible`, then you can run `foo` in both modes (like any other effect that does not use `interruptible`/`uninterruptible`).

That's why GHC moved from `block`/`unblock` to `mask` in 2010. You can dig it up.

Reply:
Even in Haskell, async exceptions can be thrown on blocking code, concurrent data structures, etc., even while inside `mask`. You can think of `interruptible` as the primitive that lets you make something interruptible deep down.

In your example, if `bar` were itself `uninterruptible`, you would have the guarantee that it would not be interrupted; but because of the new `bar.interruptible`, it would introduce left and right "edges" on which interruption could happen. So `bar` could start execution, or the fiber could be interrupted right before, or right after `bar`, because of the enclosing `bar.interruptible`. 

If you want the semantic you mention, it's now codified in `uninterruptibleMask`:

```scala
def foo = ZIO.uninterruptibleMask(restore => ... restore(bar) ...)
```

This semantic is convenient and desired in some cases (e.g. implementation of `ensuring`). However, it's not a good default for `interruptible`, because it leads to a lack of a guarantee that some inner effect can actually be interrupted.

Basically I think both semantics are required (`interruptible` and `uninterruptibleMask`). But it just so happens the former is more "primitive", as the latter is currently expressed solely in terms of the former (and a check on the current interruptibility status). You cannot express `interruptible` in terms of `uninterruptibleMask`, which is a strong indication of which one is more primitive / foundational.

Reply:
Yes, it is a primitive, but it is of no use for end-users because it is not modular (it does not respect the outer context). A user-facing `interruptible` should be a less verbose shortcut for one-level `uninterruptibleMask`'s `restore`.

Haskell has three levels of interruptibility instead of two because its async exceptions are not IO-specific and are delivered on memory allocations. And there is a stronger version of `mask` (`uninterruptibleMask`) for that extra level.

Reply:
> Yes, it is a primitive, but it is of no use for end-users because it is not modular

I strongly disagree. You can see it already used in the implementation of `bracket`, etc. It's widely useful to take an effect and ensure it can be interrupted, regardless of its embedding into uninterruptible effects. Users can utilize it to give lower-level building blocks the same status as all blocking IO, allocation, etc., receive in Haskell.

I also don't see how to make `uninterruptibleMask` with a simpler API. The problem is the user has to denote which section they want to "restore". So that means the operation needs to pass the restore function to the caller, so they can use it to restore interruptibility status in whatever region they are interested in. 

At best, you could call it `uninterruptible { restore => ... }` instead of `uninterruptibleMask { restore => ... }`. But explicit denotation of region is necessary, because if you just do `uninterruptibleMask { restore => restore(effect) }` it is the same as doing nothing at all, i.e. it just replicates the parent's interruptibility status to the child.

Reply:
The problem with this approach is that when I write
```
// step1: IO[Nothing, Unit]
// step2: IO[Nothing, Unit]
def stateTransition = (for {
  _ <- step1
  _ <- step2
} yield ()).uninterruptible
```
I no longer have any guarantee that `step2` get's executed fully after `step1` succeeds. I would have to look inside `step2` implementation to confirm that it does not use the `uninterruptible`+`interruptible` combo, all the way through the call graph.

Why would `bracket` need to *force* `use(a)` to be interruptible? It may contain no blocking code at all (e.g. only memory allocations). So you have an innocent (say, FFI-like):
```scala
def foo: IO[Nothing, Double] = bracket(allocateInMemory)(_.free)(calculate)
```
but when you use it inside `uninterruptible`, it suddenly interrupts. This is against local reasoning.

Reply:
> . I would have to look inside step2 implementation to confirm that it does not use the uninterruptible+interruptible combo, all the way through the call graph.

That is true. Use of `interruptible` is incompatible with local reasoning. However, that's already true in Haskell; and arguably it's already true in the case of defects—what happens if `step1` has a defect? Then despite the fact that you flatMap it with `step2`, `step2` will never execute.

Basically if you don't trust an effect, you should `run` it, and deal with the possibility it failed for some reason (including interruption), because that will always work correctly.

Anyway, even if we were to remove `interruptible` as a user-land primitive in favor of `uninterruptibleMask`, this would not really solve the problem if we follow suit with Haskell and make primitive operations like `sleep` interruptible (which @regiskuckaertz suggested and which I think is generally a good idea). If we have that power in ZIO, why shouldn't users have that power? After all, ZIO is not the only FFI, users will import their own foreign code and wish to make it "always interruptible".

Reply:
Btw I am open to changing semantics of `bracket` and making `use` interruption "inheritable from parent". Maybe that's a good choice here. But it's separate from the above question of whether or not to expose `interruptible` in user-land code, or to only expose `uninterruptibleMask`.

Reply:
I have no problem with exposing it, given the appropriate documentation and warning signs (and scary name, something like `forceInterruptible`). Using it in `bracket*` is definitely not appropriate (it effectively promotes `use(a)` to the class of operations that are always interruptible), so it's not self-evident that ZIO needs that power itself. As for having that class at all, I am not so sure. Many blocking operations may become non-blocking when you known something about the context (e.g. flushing a file on in-memory filesystem). Writing them using `uninterruptibleMask` actually provides more control, as you can always wrap them in `forceInterruptible` when you need it.

A simpler version of `uninterruptibleMask` (that would cover most usage patterns) could be implemented if we broaden interruptibility status to `Interruptible`, `Masked` and `Uninterruptible`:
```
uninterruptible: Interruptible -> Masked, Masked -> Uninterruptible, Uninterruptible -> Uninterruptible
interruptible (restore?): Interruptible -> Interruptible, Masked -> Interruptible, Uninterruptible -> Uninterruptible
forceInterruptible: Interruptible -> Interruptible, Masked -> Interruptible, Uninterruptible -> Interruptible
```

Reply:
I changed `bracket` to not use it. But I do want to use `interruptible` on all semantically blocking operations in ZIO (as per Haskell). Maybe separate pull request.

Reply:
It does not really force interruptibility, though. Remember that if `foo` is uninterruptible, then `interruptible(foo)` does not make it interruptible. It only adds interruptibility at the edges. And arguably, that can be changed to satisfy the two annihilation laws I mentioned:

```
interruptible(uninterruptible(foo)) == uninterruptible(foo)
uninterruptible(interruptible(foo)) == interruptible(foo)
```

What `interruptible` does is — if and only if `foo` has **no** interruptibility preferences — it makes it interruptible within an uninterruptible block. If it already has interruptibility preferences, then those will be respected.

`uninterruptibleMask` makes those interruptibility preferences inherited.

Reply:
The bottom line is that using `uninterruptibleMask` in `foo` provides the caller with the ability to control interruptibility of `foo` (*regardless* of what part of `foo` is written as potentially interruptible - the whole `foo` or a tiny `restore(bar)` section, that's an implementation detail), and using `uninterruptible`+`interruptible` does not - callee forces that on the caller (and it is not seen in the types).

Haskell's interruptions are actually exceptions. If `ZIO.sleep` had `IO[Interrupted, Unit]` result type, I would have less of a problem with it.